node/core/backing/src/lib.rs

```diff
@@ -0,0 +1,1149 @@
+
+		self.tx_from.send(FromJob::RuntimeApiMessage(RuntimeApiMessage::Request(
+				self.parent.clone(),
+				RuntimeApiRequest::ValidationCode(id, parent, parablock, tx),
+			)
+		)).await?;
+
+		Ok(rx.await?)
+	}
+
+	/// Request a `SigningContext` from the `RuntimeApi`.
+	async fn request_signing_context(&mut self) -> Result<SigningContext, Error> {
+		let (tx, rx) = oneshot::channel();
+
+		self.tx_from.send(FromJob::RuntimeApiMessage(RuntimeApiMessage::Request(
+				self.parent.clone(),
+				RuntimeApiRequest::SigningContext(tx),
+			)
+		)).await?;
+
+		Ok(rx.await?)
+	}
+
+	/// Request `HeadData` for some `ParaId` from `RuntimeApi`.
+	async fn request_head_data(&mut self, id: ParaId) -> Result<HeadData, Error> {
+		let (tx, rx) = oneshot::channel();
+
+		self.tx_from.send(FromJob::RuntimeApiMessage(RuntimeApiMessage::Request(
+				self.parent.clone(),
+				RuntimeApiRequest::HeadData(id, tx),
+			)
+		)).await?;
+
+		Ok(rx.await?)
+	}
+
+	/// Logic that runs on startup.
+	async fn on_startup(&mut self) -> Result<(), Error> {
+		let validators = self.request_validators().await?;
+	
+		let roster = self.request_validator_groups().await?; 
+
+		self.table_context.key = signing_key(&validators[..], &self.keystore);
+
+		for assignment in roster.scheduled {
+			self.table_context.groups.insert(
+				assignment.para_id,
+				roster.validator_groups[assignment.group_idx.0 as usize]
+				.iter()
+				.map(|idx| validators[*idx as usize].clone())
+				.collect()
+			);
+		}
+
+		if let Some(ref key) = self.table_context.key {
+			for (para_id, group) in self.table_context.groups.iter() {
+				if group.contains(&key.as_ref().public()) {
+					self.assignment = *para_id;
+				}
+			}
+		}
+
+		self.validation_code = self.request_validation_code(self.assignment, 0, None).await?;
+		self.head_data = self.request_head_data(self.assignment).await?;
+		self.table_context.signing_context = self.request_signing_context().await?;
+		self.table_context.validators = validators;
+
+		Ok(())
+	}
+
+	/// Run asynchronously.
+	async fn run(mut self) -> Result<(), Error> {
+		self.on_startup().await?;
+
+		while let Some(msg) = self.rx_to.next().await {
+			match msg {
+				ToJob::CandidateBacking(msg) => {
+					self.process_msg(msg).await?;
+				}
+				_ => break,
+			}
+		}
+
+		Ok(())
+	}
+
+	/// Validate the candidate that is requested to be `Second`ed and distribute validation result.
+	async fn validate_and_second(&mut self, candidate: AbridgedCandidateReceipt) -> Result<(), Error> {
+		let statement = match self.spawn_validation_work(candidate.clone()).await {
+			Ok(()) => {
+				// make PoV available for later distribution. Send data to the availability
+				// store to keep. Sign and dispatch `valid` statement to network if we
+				// have not seconded the given candidate.
+				Statement::Seconded(candidate)
+			}
+
+			// sign and dispatch `invalid` statement to network.
+			Err(_) => Statement::Invalid(candidate.hash()),
```

Comment:
A couple notes:

Is `Err` here meant to mean that there was an error validating the candidate (or communicating with the overseer, or...) or that the candidate was invalid? I would expect these meanings:
  - Ok(true): We validated successfully and the candidate was valid.
  - Ok(false): We validated successfully and the candidate was invalid.
  - Err(_): There was an error preventing us from validating. Could be an overseer error, a client error, or anything else.

In the `Ok(true)` case we should clearly be dispatching a `Statement::Seconded`.

In the `Ok(false)` case we shouldn't do anything except inform the `CandidateSelection` subsystem that the candidate was invalid. We definitely shouldn't issue a statement if nobody else has seconded the candidate, as there is no reason to mark invalid something that nobody else cares about (this is an unbounded set anyway). Technically, we could issue a `Statement::Invalid` if we observe that someone else has seconded the candidate. But it seems simpler to issue nothing, and then if we receive a `Seconded` message, we can assume that the `CandidateValidation` subsystem will cache results. Or something similar.
 
And then in the error case, I don't see any reasonable course of action other than issuing a warning and continuing.

Reply:
The `Error` type wraps all errors that may occur and so if the validation has failed it would be `Err(Error::ValidationFailed(ValidationFailed)`, so yes, here we have to match on the error type. tbh, using `Ok(bool)` seems a little less readable here.

Reply:
Still, I don't think that validation producing a negative result is an error. It is a success with a negative return type. A new enum in `primitives` would be fine for that as well.